salt/orch/prepare-product-migration.sls

```diff
@@ -0,0 +1,104 @@
+{#- Get a list of nodes seem to be down or unresponsive #}
+{#- This sends a "are you still there?" message to all #}
+{#- the nodes and wait for a response, so it takes some time. #}
+
+{%- set default_batch = salt['pillar.get']('default_batch', 5) %}
+{#- tx_update_migration_available needs to be last, as it is still used as a selector til reboot #}
+{%- set migration_grains = ['tx_update_migration_notes', 'tx_update_migration_newversion', 'migration_in_progress', 'tx_update_migration_available'] %}
+
+{%- set nodes_down = salt.saltutil.runner('manage.down') %}
+{%- if nodes_down|length >= 1 %}
+# {{ nodes_down|join(',') }} seem to be down: skipping
+  {%- do salt.caasp_log.debug('CaaS: nodes "%s" seem to be down: ignored', nodes_down|join(',')) %}
+  {%- set is_responsive_node_tgt = 'not L@' + nodes_down|join(',') %}
+{%- else %}
+# all nodes seem to be up
+  {%- do salt.caasp_log.debug('CaaS: all nodes seem to be up') %}
+  {#- we cannot leave this empty (it would produce many " and <empty>" targets) #}
+  {%- set is_responsive_node_tgt = '*' %}
+{%- endif %}
+
+{%- set is_migrateable = is_responsive_node_tgt + ' and G@tx_update_migration_available:true' %}
+
+{%- set is_migrateable_admin = 'G@tx_update_migration_available:true and G@roles:admin' %}
+
+{%- set nodes_up = salt.saltutil.runner('manage.up')|reject("equalto", "ca") %}
+
+{% for node in nodes_up %}
+  # Maybe using the cache here might lead to problems, if it is out of date.
+  {%- set migration_grain = salt['saltutil.runner']('cache.grains', tgt=node)[node].get('tx_update_migration_available', false) %}
+  {%- do salt.caasp_log.debug('Checking {0} for migration: {1}'.format(node, migration_grain)) %}
+  {% if not migration_grain %}
+    {% do salt.test.exception('Not all nodes are migratable: {0} is missing'.format(node)) %}
```

Comment:
So all the nodes that are up must be _migrateable_, or otherwise the migration will be aborted, right?. Imagine the following scenario:

1. `nodes_up=[1,2,3,4]` and `nodes_down=[666]`]
2. the user runs this orchestration
3. `nodes_up=[1,2,3,4]` are not migrateable anymore
4. `666` comes back to life
5. the user wants to run the migration again, for migrating `666`
6. the mix of migrateable (`666`) and non-mirateable (`[1,2,3,4]`) nodes abort the migration

I guess the same would happen with new nodes, right?

Reply:
@inercia True - I think bigger users often have nodes offline, so that will very likely happen.

In that case we could just switch `salt.test.exception` with `salt.caasp_log.info` to have some information in the logs about nodes missing the `migration` grain, and just proceed with the others.

Reply:
agree, right now in the update handling it's similar i think. if a single node hasn't been updated yet, you can still trigger that later